Update save my info section

cesarcosta99 opened the pull request:
Fixes 1505-gh-Automattic/woopay

#### Changes proposed in this Pull Request

Updates save my info section at checkout:

- Applies new design to both classic and Blocks checkout.
- Change its copies.
- Update tests to match new copies and the way interactions take place.

> **Note**
> 1. The copy next to the checkbox is still under discussion.
> 2. We don't know what's the link for the Learn more text yet.
>
> Check the PRD comments (link in the issue).

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

Before running the tests, make sure you have Stripe events being delivered to your local WCPay server. That way the WooPay user can be correctly created.

**Run the following tests on both classic and Blocks checkouts.**

**Test the WooPay user is created**

1. Go to the store and add a product.
2. Go to checkout and fill in the form, type an email that does not belong to a WooPay user and a valid US phone.
3. The save my info section must be visible.
4. Check the UI matches the design. There will be little distinctions between classic and Blocks, but overall they should look the same.
5. Confirm the UI looks good in the mobile version.
6. Tick the checkbox and note the phone value has been copied over the phone inner input.
7. Place the order.
8. Check the WooPay user has been created.

**Test the save my info section is NOT displayed**

1. Go to the store and add a product.
2. Go to checkout and fill in the form, type an email that belongs to a WooPay user and a valid US phone.
3. The save my info section must not be visible.
4. Place the order.
5. Confirm the WooPay user was not created.

**Test the tooltip functionality**
1. Go to the store and add a product.
2. Go to checkout and fill in the form, type an email that does not belong to a WooPay user.
3. Go to the save my info section and pass the mouse over the `i` icon, next to the WooPay logo.
4. A tooltip must show up when you hover the mouse in the icon, and it should continue visible when you move the mouse to within the tooltip.

<table>
<tr>
<th>Classic checkout</th>
<th>Blocks checkout</th>
</tr>
<tr>
<td>
<img width="332" alt="image" src="https://user-images.githubusercontent.com/10233985/208112862-a6c8c22e-0926-4ac1-a25d-f2eb7669920d.png">
</td>
<td>
<img width="492" alt="image" src="https://user-images.githubusercontent.com/10233985/208112942-994a88bb-678f-4dac-8949-82f0b049a240.png">
</td>
</tr>
</table>

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

cesarcosta99 committed:
Add new elements to save details header

cesarcosta99 committed:
Update save section agreement copy

cesarcosta99 committed:
Update save section information copy

cesarcosta99 committed:
Remove save section input label

cesarcosta99 committed:
Update save section content style

cesarcosta99 committed:
Increase margin on checkbox

cesarcosta99 committed:
Create info icon flyout

cesarcosta99 committed:
Add transition on save info section

cesarcosta99 committed:
Fix a few mobile style properties

cesarcosta99 committed:
Adjust save info tests to cover new UI

cesarcosta99 committed:
Fix save info section test

cesarcosta99 committed:
Add changelog entry

cesarcosta99 committed:
Fix tooltip SVG icon color

cesarcosta99 committed:
Tweak save info section style

cesarcosta99 committed:
Fix phone input on save info section

cesarcosta99 committed:
Update save my info implementation on Blocks checkout

cesarcosta99 committed:
Update checkbox style on Blocks checkout

cesarcosta99 committed:
Merge branch 'develop' into update/1505-save-my-info-section

cesarcosta99 commented on client/components/platform-checkout/save-user/checkout-page-save-user.js:
```diff
@@ -153,36 +167,107 @@ const CheckoutPageSaveUser = ( { isBlocksCheckout } ) => {
+						onMouseOver={ setInfoFlyoutVisible }
+						onMouseOut={ setInfoFlyoutNotVisible }
+					/>
+					<div
```
There is the [flyout component](https://developer.wordpress.org/block-editor/reference-guides/components/flyout/) from the WP components, however the package version we use does not include it.

I thought about updating the version of the package, but I've seen some deprecated properties and didn't want to go through the risk of breaking things visually. It'd be hard to catch bugs if such thing happened.

cesarcosta99 commented on client/components/platform-checkout/save-user/container.js:
```diff
@@ -0,0 +1,32 @@
+ */
+import { __ } from '@wordpress/i18n';
+
+const Container = ( { children, isBlocksCheckout } ) => {
```
Mayisha shared her findings on 1447-gh-Automattic/woopay#issuecomment-1339026618 regarding the approach we should take when developing this.

I went a bit further and found out that it's not just the `FormStep` that's not exported, but we can't import anything from `@woocommerce/base-components/cart-checkout` as it would not be found.

>We can copy that existing component and make out own wrapper to render our remember me section as a stepped block but doing so would create UI mismatch if that component gets changed in any future version of the blocks plugin.

I went down on this approach for two reasons:

- We already rely on Blocks checkout classes for this feature [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/platform-checkout/index.js#L25-L32).
- The other possibility I see is integrating it via the Blocks plugin itself, like it's mentioned there. However, IMO, I don't see as a viable option for a small feature like this. This feature has changed a lot, and I don't think the required effort for implementing such approach worths this long shot.

cesarcosta99 commented on client/settings/phone-input/index.tsx:
```diff
@@ -96,6 +95,12 @@ const PhoneNumberInput = ( {
 		};
 	}, [ onValueChange, onValidationChange ] );
 
+	useEffect( () => {
+		if ( inputInstance && inputRef.current ) {
+			onValidationChange( inputInstance.isValidNumber() );
+		}
+	}, [ value, inputInstance, inputRef, onValidationChange ] );
+
```
Had to move this `onValidationChange` call to a new `useEffect` that included `value` in the dependency array because before this component was rendered after the checkbox was clicked – which means the validation was done at the mount.

Now that there is an animation, the component is always there and needs to rely on the `value` change to check the initial validation.

cesarcosta99 committed:
Tweak save info section style

rafaelzaleski reviewed (commented):
When we copy over the number to the sign-up field, we need to either clean-up any non-numerical character or trigger an event in this field that will perform it's internal validation (`change` maybe?)

I tested with the phone `+1.646.532.9563` in the checkout field and the dots were copied over to the sign-up field. If a click in the sign-up field and remove one dot, the others are removed automatically. If I don't touch the sign-up field I'm able to submit the checkout with dots in the sign-up phone field.

rafaelzaleski reviewed (commented):
Overall it looks good and tests well. I left a comment about the validation when the phone number is copied from the WC checkout field.

I also think we should use relative units (percentages, rem, etc.) in CSS since the plugin is used with a huge variety of themes and it'd reduce the chances of it looking bad. What do you think?

cesarcosta99 committed:
Clean up phone value when copying it

cesarcosta99 committed:
Change size units from px to rem

cesarcosta99 committed:
Merge branch 'develop' into update/1505-save-my-info-section

cesarcosta99 commented:
@rafaelzaleski thanks for reviewing. I've addressed your concerns in the last commits, could you please give it another review?

>I also think we should use relative units in CSS

I agree on this and changed it, however I left a few things out of this change:
- Borders, because they won't affect other elements and the general experience.
- The places where I've used the `$gap` variables, because ideally we should aim to keep a standard UI experience with default padding/margin etc. They're still in px, and I think we should change them as well, in the package, not exactly here.

cesarcosta99 committed:
Update learn more link

rafaelzaleski reviewed (commented):
@cesarcosta99, is there a way to add the class `wc-block-components-text-input` to `PhoneNumberInputWrapper` if we are in the checkout block? 

I tested with another theme and the input didn't look good without this class.

<img width="725" alt="Screenshot 2022-12-19 at 17 50 08" src="https://user-images.githubusercontent.com/4153905/208521165-7041773e-7df1-427e-95f4-e4915d9bbbc6.png">

cesarcosta99 committed:
Fix input style in Blocks checkout

On a different theme, the input may look strange. We're adding a class to try to make it look as it should.

cesarcosta99 committed:
Fix some theme incompatibilities

cesarcosta99 committed:
Enforce padding and margin values to avoid more theme incompatibilities

cesarcosta99 committed:
Merge branch 'develop' into update/1505-save-my-info-section

rafaelzaleski reviewed (approved):
Looks good and tests well! 🚢
